src/components/backend-ai-resource-broker.ts

```diff
@@ -671,6 +646,54 @@ export default class BackendAiResourceBroker extends BackendAIPage {
       this.available_slot = remaining_sg_slot;
       this.used_slot_percent = used_slot_percent;
       this.used_resource_group_slot_percent = used_resource_group_slot_percent;
+
+      let enqueueSession = false;
+      if (globalThis.backendaiclient._config.always_enqueue_compute_session === true) {
+        enqueueSession = true;
+      }
+      if (response.presets) {
+        const presets = response.presets;
+        const availablePresets: any = [];
+        presets.forEach((item) => {
+          if (item.allocatable === true) {
+            for (const [slotKey, slotName] of Object.entries(slotList)) {
+              if (slotKey in item.resource_slots) {
+                item[slotName] = item.resource_slots[slotKey];
+              }
+            }
+          } else if (enqueueSession) {
+            // Even if allocatable is false, if enqueueSession is true,
+            // allocatable is determined based on when no resources are allocated.
+            item.allocatable = true;
+            for (const [slotKey, slotName] of Object.entries(slotList)) {
+              if (slotKey in item.resource_slots && slotName in total_resource_group_slot) {
+                if (item.resource_slots[slotKey] <= total_resource_group_slot[slotName]) {
+                  item[slotName] = item.resource_slots[slotKey];
+                } else {
+                  item.allocatable = false;
+                  break;
+                }
+              }
+            }
+          }
+
+          // Change to binary unit
+          item.mem = globalThis.backendaiclient.utils.changeBinaryUnit(item.resource_slots.mem, 'g');
+          if (item.shared_memory) {
+            item.shmem = globalThis.backendaiclient.utils.changeBinaryUnit(item.shared_memory, 'g');
+          } else {
+            item.shmem = null;
+          }
+          if (item.allocatable) {
+            availablePresets.push(item);
+          }
+        });
```

Comment:
You can divide this part into `map` and `filter`. This is more easy to read and easy to handle type.
```js
const availablePresets  = presets.map(item => {
  //modify item
  return item
}).filter(item => item.allocatable)

```